torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,223 @@
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(
+                child_sync_inserter.final().begin(),
+                child_sync_inserter.final().end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    // This level of the nested for-loop may not exist in the kernel.
+    // However, subsequent levels can exist, so we handle the body of the
+    // for-loop first.
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
+      // Determine if any smem TV is written to at beginning of the for-loop
+      // and whether that smem TV is read from at the end of the for-loop
+      size_t isize = intersection_size(initial_, final_);
+      // Insert new SyncThreads at end of for-loop to prevent WAR race condition
+      if (isize > 0 &&
+          fl->body().exprs().back()->getExprType().value() != ExprType::Sync &&
+          !is_last_op_sync_) {
+        // std::cout << "WAR race detected; Add Sync" << std::endl;
+        has_war_hazard_sync_ = true;
+        fl->body().push_back(new kir::Sync(true));
+      }
+    }
+  }
+
+  size_t intersection_size(
+      std::unordered_set<const TensorView*>& left,
+      std::unordered_set<const TensorView*>& right) {
+    size_t result = 0;
+    for (auto item : left) {
+      if (right.find(item) != right.end()) {
+        ++result;
+      }
+    }
+    return result;
+  }
+
+  void hasOutputSmemExpr(
+      Expr* expr,
+      std::unordered_set<const TensorView*>& set) {
+    for (auto out : expr->outputs()) {
+      if (ir_utils::isTV(out)) {
+        auto tv = out->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+  void hasInputSmemExpr(
+      Expr* expr,
+      std::unordered_set<const TensorView*>& set) {
+    for (auto inp : expr->inputs()) {
+      if (ir_utils::isTV(inp)) {
+        auto tv = inp->as<TensorView>();
+        if (tv->getMemoryType() == MemoryType::Shared) {
+          set.insert(tv);
+        }
+      }
+    }
+  }
+
+ private:
+  // Track Shared Memory Inputs (Reads) for parent for-loop
+  std::unordered_set<const TensorView*> all_smem_inputs_;
+
+  // Track Shared Memory Outputs (Writes) for parent for-loop
+  std::unordered_set<const TensorView*> all_smem_outputs_;
+
+  // Shared Memory Writes at beginning of the for-loop
+  // before first SyncThreads
+  std::unordered_set<const TensorView*> initial_;
+
+  // Shared Memory Reads at end of the for-loop
+  // Cleared after each SyncThreads
+  std::unordered_set<const TensorView*> final_;
+
+  // Track first sync found in for-loop
+  bool initial_sync_ = false;
+
+  // Track sync was inserted for war hazard
+  bool has_war_hazard_sync_ = false;
+};
+
+} // namespace
+
+std::vector<Expr*> insertThreadSynchronization(
+    Fusion* fusion,
+    const std::vector<Expr*>& exprs) {
+  FusionGuard fg(fusion);
```

Comment:
Please add a perf scope marker (`FUSER_PERF_SCOPE("insertThreadSynchronization");). It should be the first thing in the function body.

Reply:
:+1: